src/async_api/tokio/mod.rs

```diff
@@ -0,0 +1,298 @@
+            }
+        }
+    }
+
+    async fn session_configuration(&self) -> Option<super::active_session::SessionConfiguration> {
+        let cfg = self.session_configuration.lock().await;
+        if cfg.proto_major > 0 {
+            Some((*cfg).clone())
+        } else {
+            None
+        }
+    }
+}
+
+struct WorkerIsTerminated(Option<Pin<Box<tokio::sync::oneshot::Receiver<()>>>>);
+impl<'a> Future for &'a mut WorkerIsTerminated {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let e = self.0.take();
+        if let Some(mut e) = e {
+            if let Poll::Pending = e.as_mut().poll(cx) {
+                self.0 = Some(e);
+                Poll::Pending
+            } else {
+                Poll::Ready(())
+            }
+        } else {
+            Poll::Ready(())
+        }
+    }
+}
+
+struct TokioWorker {
+    terminated: WorkerIsTerminated,
+    sub_channel: Arc<SubChannel>,
+}
+
+#[async_trait::async_trait]
+impl Worker for TokioWorker {
+    async fn wait_for_shutdown(&mut self) -> () {
+        let r = &mut self.terminated;
+        let _ = r.await;
+        ()
+    }
+
+    async fn read_single_request<'a, 'b>(
+        &mut self,
+        buffer: &'b mut [u8],
+    ) -> Option<io::Result<super::Request<'b>>> {
+        let sub_channel = self.sub_channel.clone();
+
+        let r = tokio::select! {
+         _ = self.wait_for_shutdown() => {
+             Ok(None)
+         }
+        result = sub_channel.do_receive(buffer) => {
+             result
+         }
+        };
+        match r {
+            Err(err) => match err.raw_os_error() {
+                // Operation interrupted. Accordingly to FUSE, this is safe to retry
+                Some(ENOENT) => None,
+                // Interrupted system call, retry
+                Some(EINTR) => None,
+                // Explicitly try again
+                Some(EAGAIN) => None,
+                // Filesystem was unmounted, quit the loop
+                Some(ENODEV) => Some(Err(err)),
+                // Unhandled error
+                _ => Some(Err(err)),
+            },
+            Ok(Some(size)) => {
+                if let Some(req) = crate::async_api::Request::new(&buffer[..size]) {
+                    Some(Ok(req))
+                } else {
+                    None
+                }
+            }
+            Ok(None) => None,
+        }
+    }
+
+    async fn sender(&self) -> Arc<dyn super::reply::ReplySender> {
+        self.sub_channel.clone()
+    }
+}
+
+pub(in crate::async_api) struct OpenedTokio {
+    channel: channel::Channel,
+}
+
+#[async_trait::async_trait]
+impl OpenedFlavor for OpenedTokio {
+    async fn spawn_run(
+        self,
+        filesystem: Arc<dyn Filesystem>,
+    ) -> io::Result<Arc<dyn SessionHandle>> {
+        let OpenedTokio { channel } = self;
+
+        let sub_channels = channel.sub_channels.clone();
```

Comment:
I think if you `drain()` the `sub_channels` instead of cloning them, their ownership can be moved into `TokioWorker` and then we could rely on `Drop` instead of needing to `.close()` them in the `Channel::drop` method

Reply:
the close is part of ensuring they stop operating -- oh if we don't close them aggressively, one of the child FD's could notice the system is shutting down. But the other children don't. 

Due to limitations in the mac fuse the libc reads are blocking ones, so we want the sockets closed to ensure they don't hang after something notices the mount is broken/wants to shut down.

Reply:
I see, but wouldn't the `terminated` signal prevent them from hanging? I probably don't understand what all it's for, but I thought that was the signal we used to notify all the children that they should exit

Reply:
I added a comment on this -- mostly its around the blocking path. Right now terminated signal could be ignored/or not respected quickly in the blocking path since we have to use the libc::read.